boards: arm: Add adafruit_kb2040 board

petejohanson opened the pull request:
Add `adafruit_kb2040` board with `sparkfun,pro-micro` header.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

yonsch commented on boards/arm/adafruit_kb2040/Kconfig.board:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2021 Pete Johanson
```
Not sure if it's important or not, but 2022 (fix everywhere)

petejohanson replied:
Fixed.

yonsch commented on boards/arm/adafruit_kb2040/adafruit_kb2040-pinctrl.dtsi:
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2021, Yonatan Schachter
```
Not sure about the rules here, but you should probably add yourself below or replace this line altogether.

yonsch commented on boards/arm/adafruit_kb2040/adafruit_kb2040.dts:
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2021 Yonatan Schachter
```
Same here

petejohanson replied:
Added a second line w/ my name and updated date.

yonsch commented on boards/arm/adafruit_kb2040/adafruit_kb2040.dts:
```diff
@@ -0,0 +1,88 @@
+};
+
+&flash0 {
+	/* 8MB of flash minus the 0x100 used for
```
This can be removed (this is an old comment from the pico which is no longer relevant)

petejohanson replied:
Done.

yonsch commented on boards/arm/adafruit_kb2040/adafruit_kb2040.dts:
```diff
@@ -0,0 +1,88 @@
+		#size-cells = <1>;
+
+		/*
+		 * Start at the beginning of usable flash, 8MB minus the
```
Where is the 2nd stage BL's partition?

petejohanson replied:
Restored.

yonsch commented on boards/arm/adafruit_kb2040/adafruit_kb2040.yaml:
```diff
@@ -0,0 +1,12 @@
+identifier: raspberrypi_pico
```
Wrong identifier/name

petejohanson replied:
Fixed.

yonsch commented on boards/arm/adafruit_kb2040/adafruit_kb2040.yaml:
```diff
@@ -0,0 +1,12 @@
+name: RaspberryPi-Pico
+type: mcu
+arch: arm
+flash: 2048
```
Shouldn't this be 8MB?

petejohanson replied:
Indeed. Fixed.

yonsch commented on boards/arm/adafruit_kb2040/adafruit_kb2040.yaml:
```diff
@@ -0,0 +1,12 @@
+  - gnuarmemb
+  - xtools
+supported:
+  - serial
```
There are many more, copy from the pico

petejohanson replied:
Updated.

yonsch commented on boards/arm/adafruit_kb2040/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+
+The Adafruit KB2040 is a small, low-cost, versatile board from
+Adafruit. It is equipped with an RP2040 SoC, an on-board RGB Neopixel,
+a USB connector, and an QT connector. The USB bootloader allows it
```
a QT connector

petejohanson replied:
Added the full name "STEMMA QT connector" to match the adafruit docs/naming.

yonsch commented on boards/arm/adafruit_kb2040/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+********
+- Dual core Arm Cortex-M0+ processor running up to 133MHz
+- 264KB on-chip SRAM
+- 2MB on-board QSPI flash with XIP capabilities
```
8MB?

petejohanson replied:
Fixed

yonsch commented on boards/arm/adafruit_kb2040/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+- Dual core Arm Cortex-M0+ processor running up to 133MHz
+- 264KB on-chip SRAM
+- 2MB on-board QSPI flash with XIP capabilities
+- 26 GPIO pins
```
Are all of them accessible on this board?

petejohanson replied:
No. Fixed.

yonsch commented on boards/arm/adafruit_kb2040/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+Supported Features
+==================
+
+The adafriut_kb2040 board configuration supports the following
```
adafriut is a typo

petejohanson replied:
Fixed

yonsch reviewed (changes requested):
A few nitpicks. Also, if the `board.cmake` file is not used (since there's currently no west configuration for it as west does not support MSC), I think it can be omitted. Regarding the `sparkfun_pro_micro` - I see there's a board of that name. Is it also a name of a standard pinout? Or does this port support both boards? Maybe the file should be renamed to `sparkfun_pro_micro_connector.dtsi`?

petejohanson commented:
> A few nitpicks. Also, if the `board.cmake` file is not used (since there's currently no west configuration for it as west does not support MSC), I think it can be omitted. Regarding the `sparkfun_pro_micro` - I see there's a board of that name. Is it also a name of a standard pinout? Or does this port support both boards? Maybe the file should be renamed to `sparkfun_pro_micro_connector.dtsi`?

It's a board, but one that's AVR, so won't ever run Zephyr. Renamed the include for completeness. We use the generic header/interconnect extensively in ZMK since there are so many keyboards that use the pro micro footprint.

Thanks for the thorough review!

petejohanson commented:
@yonsch Found and fixed the remaining CI issues.

henrikbrixandersen reviewed (changes requested):
Looks good. However, please split out the commits. The new DTS binding, which is not specific to the board, should go into its own commit. Same goes for the change to the ADC test.

petejohanson committed:
dts: bindings: gpio: Add SparkFun Pro Micro header

Add SparkFun Pro Micro header connector that is implemented by many
other controllers. This allows hardware with compatible headers to
define the related GPIOs and peripherals.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

petejohanson committed:
dts: bindings: vendor-prefixes: add SparkFun

Add SparkFun (https://www.sparkfun.com/) vendor prefix.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

petejohanson committed:
boards: arm: Add adafruit_kb2040 board

Add `adafruit_kb2040` board with `sparkfun,pro-micro` header.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

petejohanson committed:
drivers: adc: Fix ADC tests for new RP2040 board.

Fix ADC tests now that we have more than one RP2040 board.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

petejohanson commented on boards/arm/adafruit_kb2040/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+.. target-notes::
+
+.. _Getting Started with Raspberry Pi Pico:
+  https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf
```
Added a link to Adafruit's guide, which includes pinouts, schematics, etc.

They don't seem to have a true dedicated "datasheet" PDF for it though.

soburi replied:
> They don't seem to have a true dedicated "datasheet" PDF for it though.

I got it. Thank you!

petejohanson commented on boards/arm/adafruit_kb2040/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+- 4 Analog inputs
+- 1 UART peripherals
+- 1 SPI controllers
+- 1 I2C controllers
```
Tweaked the description a bit here. LMK if the new version works for you.

petejohanson commented:
@henrikbrixandersen Split the commits, including an extra one just for the new vendor prefix for SparkFun as well.

henrikbrixandersen commented on boards/arm/adafruit_kb2040/adafruit_kb2040_defconfig:
```diff
@@ -0,0 +1,27 @@
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=125000000
+
+# Enable GPIO
+CONFIG_GPIO=y
```
If the board requires GPIO in order to _boot_ and use the console (e.g. if it has a GPIO controlled regulator, which needs to be turned on), GPIO should be enabled by default. Otherwise not. And yet, we do have a lot of boards enabling GPIO by default where it is not needed.
